catalog/app/utils/PFSCookieManager.tsx

```diff
@@ -0,0 +1,114 @@
+import * as dateFns from 'date-fns'
+import invariant from 'invariant'
+import * as React from 'react'
+import * as redux from 'react-redux'
+import { createSelector } from 'reselect'
+
+import cfg from 'constants/config'
+import { tokens as tokensSelector } from 'containers/Auth/selectors'
+import { useApi } from 'utils/APIConnector'
+
+type Ensure = (ttl: number, histeresis: number) => Promise<void>
+
+const Ctx = React.createContext<Ensure | null>(null)
+
+const selectToken = createSelector(
+  tokensSelector,
+  (tokens) => tokens?.token as string | undefined,
+)
+
+interface State {
+  token: string | undefined
+  expires: Date
+  promise: Promise<void>
+}
+
+function needsRefresh(
+  state: State | undefined,
+  params: {
+    token: string | undefined
+    expires: Date
+    histeresis: number
+  },
+): boolean {
+  if (!state) return true
+  if (state.token !== params.token) return true
+  if (state.expires.getTime() + params.histeresis < params.expires.getTime()) return true
+  return false
+}
+
+export function PFSCookieManager({ children }: React.PropsWithChildren<{}>) {
+  const stateRef = React.useRef<State>()
+
+  const req = useApi()
+
+  const setPFSCookie = React.useCallback(
```

Comment:
Maybe, `requestPFSCookie`, so not to confuse with `React.setState` and highlight that this function is not cheap

Reply:
I think, you can make this callback an ordinary function (more likely two functions), but probably it's not worth it, and not much time left

Reply:
if i split them, nothing except the actual request call will remain in the "ordinary" function